src/integrations/hapiApollo.ts

```diff
@@ -1,52 +1,145 @@
 import * as hapi from 'hapi';
 import * as graphql from 'graphql';
+import * as GraphiQL from '../modules/renderGraphiQL';
 import { runQuery } from '../core/runQuery';
+import ApolloOptions from './apolloOptions';
 
 
 export interface IRegister {
     (server: hapi.Server, options: any, next: any): void;
     attributes?: any;
 }
 
-export interface HapiApolloOptions {
-    schema: graphql.GraphQLSchema;
-    formatError?: Function;
-    rootValue?: any;
-    context?: any;
-    logFunction?: Function;
+export interface HAPIOptionsFunction {
+  (req?: hapi.Request): ApolloOptions | Promise<ApolloOptions>;
 }
 
-export class HapiApollo {
+export class ApolloHAPI {
     constructor() {
         this.register.attributes = {
             name: 'graphql',
             version: '0.0.1',
         };
     }
 
-    public register: IRegister = (server: hapi.Server, options: HapiApolloOptions, next) => {
+    public register: IRegister = (server: hapi.Server, options: ApolloOptions | HAPIOptionsFunction, next) => {
         server.route({
-            method: 'GET',
-            path: '/test',
-            handler: (request, reply) => {
-                reply('test passed');
+            method: 'POST',
+            path: '/',
+            handler: async (request, reply) => {
+              let optionsObject: ApolloOptions;
+              if (isOptionsFunction(options)) {
+                try {
+                  optionsObject = await options(request);
+                } catch (e) {
+                  reply(`Invalid options provided to ApolloServer: ${e.message}`).code(500);
+                }
+              } else {
+                optionsObject = options;
+              }
+
+              if (!request.payload) {
+                reply('POST body missing.').code(500);
+                return;
+              }
+
+              const responses = await processQuery(request.payload, optionsObject);
+
+              if (responses.length > 1) {
+                reply(responses);
+              } else {
+                const gqlResponse = responses[0];
+                if (gqlResponse.errors && typeof gqlResponse.data === 'undefined') {
+                  reply(gqlResponse).code(400);
```

Comment:
I'm starting to wonder whether we should have gotten rid of HTTP error codes altogether. What do you think?